usb: device_next: align CDC ACM UART with Interrupt-driven UART API

jfischer-no opened the pull request:
Align CDC ACM UART with Interrupt-driven UART API behavior description.
Use the same flags in uart_irq_rx_ready(), uart_irq_tx_ready(), and
uart_irq_is_pending(), which are updated after each uart_irq_update()
call. Allow TX FIFO to be filled if there is space.

Add CDC ACM virtual UART Interrupt-driven API documentation.

- [x] poll_in/poll_out needs also be aligned with legacy CDC ACM UART behavior
- [x] poll_in/poll_out documentation

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/76642

jfischer-no committed:
doc: usb: move "Built-in functions" down  and increase the heading

Move "Built-in functions" section down and increase the heading.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: align CDC ACM UART with Interrupt-driven UART API

Align CDC ACM UART with Interrupt-driven UART API behavior description.
Use the same flags in uart_irq_rx_ready(), uart_irq_tx_ready(), and
uart_irq_is_pending(), which are updated after each uart_irq_update()
call. Allow TX FIFO to be filled if there is space.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: usb: describe CDC ACM virtual UART IRQ API behavior

In general, it mirrors what is described in Interrupt-driven API
documentation, but here a little more explicitly and with a simplified
example, so that we can finally chisel it into the stone.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: align CDC ACM UART poll_out with legacy implementation

Apply changes in commit c152e0980cf6
("usb: device: cdc_acm: block in uart_poll_out() routine")
to the new CDC ACM UART poll_out implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow fifo_fill and poll_out be used simultaneously

As it is still accepted practice, allow fifo_fill and poll_out to be
used simultaneously. The lock around fifo_fill was already in place.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: usb: add note about CDC ACM virtual UART polling API behavior

Add note about CDC ACM virtual UART polling API behavior.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented:
Commit "usb: device_next: allow fifo_fill and poll_out be used simultaneously" in this PR fixes #76642

jfischer-no committed:
usb: device_next: use delayable work for TX FIFO in CDC ACM

Use delayable work to reduce CPU load when there is no transfer flow in
the host direction. This also improves the performance of poll out, as
introduced in commit commit fed6bde78852
("usb: device: cdc_acm: send more than 1 byte in poll out")
for the legacy CDC ACM implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: limit CDC ACM OUT transfer size to the current MPS

When the controller is connected to a full speed bus, regardless of
whether the controller supports high speed or not, the transfer size for
the bulk OUT endpoint should be equal to the MPS in the current
configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented:
There is conflict in `doc/connectivity/usb/device_next/usb_device.rst` that has to be fixed.

aescolar commented:
@jfischer-no rebase needed
